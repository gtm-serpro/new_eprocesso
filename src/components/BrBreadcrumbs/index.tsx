import CustomTag from "@/common/CustomTag";
import { memo } from "react";

export type Breadcrumb = {
  label: string;
  isHome?: boolean;
  href?: string;
  onClick?: React.MouseEventHandler<HTMLElement>;
  active?: boolean;
};

export type BreadcrumbsProps = {
  crumbs?: Breadcrumb[];
};

const handleClick = (onClick: Breadcrumb["onClick"]) => (e: React.MouseEvent<HTMLElement>) => {
  if (onClick) {
    e.preventDefault();
    e.stopPropagation();
    onClick(e);
  }
};

const HomeCrumb = memo(({ label, href, onClick }: Breadcrumb) => {
  return (
    <li className="crumb home mr-0">
      <span className="sr-only">{label}</span>
      <a className="br-button circle small" onClick={handleClick(onClick)} href={href}>
        <i className="fas fa-home" />
      </a>
    </li>
  );
});

const RouteCrumb = memo(({ label, href, active, onClick }: Breadcrumb) => {
  return (
    <li className="crumb">
      <i className="text-center icon fas fa-chevron-right" />
      <CustomTag tagName={active ? null : "a"} onClick={handleClick(onClick)} href={href}>
        <span>{label}</span>
      </CustomTag>
    </li>
  );
});

const Crumb = memo(({ isHome, ...rest }: Breadcrumb) => {
  return isHome ? <HomeCrumb {...rest} /> : <RouteCrumb {...rest} />;
});

const BrBreadcrumbs = ({ crumbs = [] }: BreadcrumbsProps) => {
  return (
    <div className="br-breadcrumb">
      <ul className="crumb-list">
        {crumbs.map((crumb, index) => (
          <Crumb key={index} {...crumb} />
        ))}
      </ul>
    </div>
  );
};

export default memo(BrBreadcrumbs);
