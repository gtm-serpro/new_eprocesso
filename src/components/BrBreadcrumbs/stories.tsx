import { Meta, Story } from "@storybook/react";
import BrBreadcrumbs from ".";

export default {
  title: "Navegação/Breadcrumbs",
  component: BrBreadcrumbs,
} as Meta;

const bc = [
  {
    isHome: true,
    label: "home",
    onClick: () => {
      alert("home");
    },
  },
  {
    label: "rota",
    href: "#",
    onClick: () => {
      alert("rota");
    },
  },
  {
    label: "subrota",
    active: true,
  },
];

export const Simples: Story = (args) => <BrBreadcrumbs {...args} />;
Simples.args = { ...Simples.args, crumbs: bc };
