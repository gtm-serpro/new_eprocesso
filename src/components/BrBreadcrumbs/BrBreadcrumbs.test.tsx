import { fireEvent, render, screen } from "@testing-library/react";
import BrBreadcrumbs, { Breadcrumb } from ".";

describe("<BrBreadcrumbs/>", () => {
  it("deveria renderizar lista vazia de crumbs se nada for informado", () => {
    const { container } = render(<BrBreadcrumbs />);
    expect(container.firstChild).toHaveClass("br-breadcrumb");
    expect(screen.getByRole("list")).toHaveClass("crumb-list");
    expect(screen.queryByRole("listitem")).toBeNull();
  });

  it("deveria renderizar lista de crumbs informadas", () => {
    const crumbs = [
      { label: "crumb-1", onClick: () => true },
      { label: "crumb-2", onClick: () => true },
      { label: "crumb-3", onClick: () => true },
    ];

    const { container } = render(<BrBreadcrumbs crumbs={crumbs} />);

    expect(container.firstChild).toHaveClass("br-breadcrumb");
    expect(screen.getByRole("list")).toHaveClass("crumb-list");

    screen.getAllByRole("listitem").forEach((item) => {
      expect(item).toHaveClass("crumb");
      expect(item.firstChild).toHaveClass("text-center icon fas fa-chevron-right");
    });

    expect(container.firstChild).toMatchSnapshot();
  });

  it("deveria renderizar lista de crumbs com 1 home", () => {
    const crumbs: Breadcrumb[] = [
      { label: "home", onClick: () => true, isHome: true },
      { label: "crumb-2", onClick: () => true },
      { label: "crumb-3", onClick: () => true },
    ];
    render(<BrBreadcrumbs crumbs={crumbs} />);

    expect(screen.getAllByRole("listitem")[0]).toHaveClass("crumb home");
  });

  it("deveria responder ao clique nas crumbs", () => {
    const handleClick = jest.fn();

    const crumbs = [
      { label: "crumb-1", onClick: handleClick },
      { label: "crumb-2", onClick: handleClick },
      { label: "crumb-3", onClick: handleClick },
    ];

    render(<BrBreadcrumbs crumbs={crumbs} />);

    fireEvent.click(screen.getByText("crumb-1"));
    fireEvent.click(screen.getByText("crumb-2"));
    fireEvent.click(screen.getByText("crumb-3"));

    expect(handleClick).toHaveBeenCalledTimes(3);
  });
});
