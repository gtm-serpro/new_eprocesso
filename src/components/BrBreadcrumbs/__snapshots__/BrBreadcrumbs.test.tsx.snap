// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BrBreadcrumbs/> deveria renderizar lista de crumbs informadas 1`] = `
<div
  class="br-breadcrumb"
>
  <ul
    class="crumb-list"
  >
    <li
      class="crumb"
    >
      <i
        class="text-center icon fas fa-chevron-right"
      />
      <a>
        <span>
          crumb-1
        </span>
      </a>
    </li>
    <li
      class="crumb"
    >
      <i
        class="text-center icon fas fa-chevron-right"
      />
      <a>
        <span>
          crumb-2
        </span>
      </a>
    </li>
    <li
      class="crumb"
    >
      <i
        class="text-center icon fas fa-chevron-right"
      />
      <a>
        <span>
          crumb-3
        </span>
      </a>
    </li>
  </ul>
</div>
`;
